Topbar component

b-kelly opened the pull request:
Stacks topbar component. Aims to achieve visual fidelity with what we have on the Stack Exchange network, while relying on Stacks for providing the bulk of the styling.

b-kelly committed:
initial commit of topbar port

b-kelly committed:
tweaks

b-kelly committed:
add ctas

b-kelly committed:
search cleanup

b-kelly committed:
move topbar out to partial, add s-topbar--notice

b-kelly committed:
cleanup

b-kelly committed:
make the topbar variants valid mixins because... reasons

b-kelly committed:
cleanup

b-kelly committed:
cleanup

b-kelly committed:
cleanup

b-kelly committed:
tweak --topbar-* to --theme-topbar-*

b-kelly committed:
remove "contained" implication

b-kelly committed:
move variables into .components

b-kelly committed:
Merge branch 'develop' into bkelly/topbar-component

b-kelly committed:
switch badge indicators to filled activity indicators

aaronshekey commented on lib/css/components/_stacks-topbar.less:
```diff
@@ -0,0 +1,366 @@
+
+    // TODO necessary? used to wrap content so the topbar stretches 100% w/ content at some value below that
+    .s-topbar--container {
+        width: 100%; // TODO note - consumers should use atomic classes to override this
```
I think by default, we can set this to be `@site-width` or `wmx12`. The non-default behavior we may see is someone adding `wmx-initial` to `.s-topbar--container`.

b-kelly replied:
Set it to `width: @s-full; max-width: 100%;`. This looks pretty good in the general case as well as on our squished examples in the docs. I'd prefer not to use `@site-width` since I'm actively trying to remove that and doing so would require me to move this component out of `static` and into `dynamic`, which kills my ability to use it in non-unified areas (SE.com)

aaronshekey commented on lib/css/components/_stacks-topbar.less:
```diff
@@ -0,0 +1,366 @@
+                left: 0;
+                background: var(--theme-topbar-item-color);
+                top: -5px;
+                transition: all ease-in-out 0.1s;
```
Nit: I try to avoid using `all` since that'll add a transition to color changes and other unpredictable things. Better to just define `transform` and `top` or whatever individual thing.

aaronshekey commented on lib/css/components/_stacks-topbar.less:
```diff
@@ -0,0 +1,366 @@
+        position: relative;
+        white-space: nowrap;
+
+        &:hover, &:focus,
```
All separate lines, please. If you've got `stylelint` installed in your editor, it should underline these as linting errors :v:

b-kelly replied:
Good call! I just installed the stylelint extension locally. An option to help enforce this would be to add it to our recommended extensions for VS Code (right click the extension in the pane, then `Add to Workspace Recommendations`).

aaronshekey commented on lib/css/components/_stacks-topbar.less:
```diff
@@ -0,0 +1,366 @@
+    margin-left: auto;
+
+    li + li {
+        margin-left: @su4;
```
`gap` can not come soon enough.

aaronshekey commented on lib/css/components/_stacks-topbar.less:
```diff
@@ -0,0 +1,366 @@
+    --theme-topbar-search-placeholder:         @black-200;
+    --theme-topbar-search-border:              @black-200;
+    --theme-topbar-search-border-focus:        @blue-300;
+    --theme-topbar-search-shadow-focus:        0 0 0 @su4 fade(@blue-500, 15%);
```
This likely still needs to be aware of `var(--theme-primary)`. Borrow the approach we used on buttons? Will `0 0 0 @su4 var(--focus-ring);` cut it?

b-kelly replied:
Good call on this as well. I was using the value on Prod, but I like yours better.

aaronshekey commented on lib/css/components/_stacks-topbar.less:
```diff
@@ -0,0 +1,366 @@
+        color: var(--theme-topbar-item-color);
+        display: inline-flex;
+        align-items: center;
+        padding: 0 10px; // TODO @su8 or @su12?
```
12, or... if you really wanna be fancy, you could `@su12 - @su2`. We've done worse.

b-kelly replied:
Imo, 12 was too much and 8 was too little. I added `(@su12 - @su2)` as suggested so we're at least not _entirely_ hardcoding the value...

aaronshekey commented on lib/css/components/_stacks-topbar.less:
```diff
@@ -0,0 +1,366 @@
+            &:after {
+                position: absolute;
+                content: '';
+                width: 18px;
```
We should put this down to `@su16` for more consistent rendering with `@Svg.Hamburger` but we'll need to double check all the rotation math.

b-kelly replied:
No math changes needed, thankfully!

aaronshekey commented on lib/css/components/_stacks-topbar.less:
```diff
@@ -0,0 +1,366 @@
+        display: flex;// none;
+        padding: 0;
+        height: 100%;
+        width: 44px;
```
Should we make this a perfect square and just go `@su48`?

b-kelly replied:
Nice catch! I'll do one better and not hardcode the width at all. Changing the items inside to 16px means that I can do `padding: 0 16px` and get the same width out, but with some flexibility if the interior size changes up any. This was a holdover from the legacy css I ~~copy/pasted~~ ported over.

aaronshekey commented on docs/product/components/partials/_topbar.html:
```diff
@@ -0,0 +1,151 @@
+                <li>
+                    <a href="#" class="s-topbar--item" title="Review queues">
+                        {% icon "ReviewQueue" %}
+                        <div class="s-activity-indicator">
```
I think we want `s-activity-indicator__danger` here.
```suggestion
                        <div class="s-activity-indicator s-activity-indicator__danger">
```

b-kelly replied:
Kinda. On production that indicator can be either blue or red, depending on if the queue "is highlighted", which is based on a couple of factors (queue length, any item is "in danger"). Since I already had one red item in the examples, I decided to keep this one blue.

b-kelly replied:
I went back and looked at the code again and for the life of me I can't find any trace of the blue (vs red) indicator. Maybe I dreamt it? I'll look over it again, but I might be describing some legacy topbar styles that are no longer in use... I'll update the indicator to be danger to match.

aaronshekey commented on docs/product/components/partials/_topbar.html:
```diff
@@ -0,0 +1,151 @@
+                        <div class="s-user-card--info">
+                            <ul class="s-user-card--awards">
+                                <li class="s-user-card--rep">3,145</li>
+                                <li class="s-award-bling s-award-bling__gold">3</li>
```
Speaking of interdependencies, we won't be able to go live with this until we add custom CSS for bling. Otherwise this'll be a regression across the network. Not news to you of course, but noting it here for when we both forget lol.

b-kelly replied:
On Core I have this working / looking ok with the current markup that we're using for the bling. I want the Stacks docs to be 100% Stacks, but the topbar component is flexible enough in this aspect that we have quite a bit of wiggle room as to what we put inside the `s-topbar--item`s.

aaronshekey commented on docs/product/components/partials/_topbar.html:
```diff
@@ -0,0 +1,151 @@
+                <li>
+                    <a href="#" class="s-topbar--item" title="Inbox">
+                        {% icon "Inbox" %}
+                        <span class="s-activity-indicator s-activity-indicator__danger">0</span>
```
```suggestion
                        <span class="s-activity-indicator">3</span>
```

aaronshekey commented on docs/product/components/topbar.html:
```diff
@@ -0,0 +1,42 @@
+<section class="stacks-section">
+    {% header "h2", "Logged in" %}
+
+    {% assign variations = "s-topbar__auto,s-topbar__dark,s-topbar__light" | split: "," %}
```
Is this here 'cause you don't wanna pass an empty class or something? `s-topbar__auto` is the same as just a plain ol' `s-topbar` yeah? Think you can just kill it.

b-kelly replied:
More or less. I show the class above the generated example right now, but that might change once the docs are actually written. In that case, we can make this an empty class or whatever really.

aaronshekey commented on lib/css/components/_stacks-topbar.less:
```diff
@@ -0,0 +1,366 @@
+    // TODO Stacks component? Badge?
+    .s-topbar--item.s-topbar--notice {
+        border-radius: @br-sm;
+        font-weight: bold;
```
We have more control these days. `700` or `800` ought to do it.

b-kelly replied:
Do you have any specific tips on when to use `bold` vs the numeric values? I don't personally have any specific sense for when one looks better than another or on what fonts these additional values are supported. Fwiw, it seems like Less is compiling this into `700`. I didn't realize it did that!

aaronshekey commented on lib/css/components/_stacks-topbar.less:
```diff
@@ -0,0 +1,366 @@
+.s-topbar .s-topbar--searchbar {
+    @inputLineHeights: @lh-sm; // ensure the line heights between the two elements match up
+
+    -webkit-transform: translate3d(0, 0, 0); // Solves an issue with Safari clipping things strangely when a popover is present in the topbar
```
Bust out that mac and see if this is still an issue?

b-kelly replied:
Tested it and it didn't seem to be an issue any longer

b-kelly committed:
Merge branch 'develop' into bkelly/topbar-component

b-kelly committed:
stylelint

b-kelly committed:
swap out atomicly built topbar for actual topbar component

b-kelly committed:
cleanup

b-kelly committed:
container width tweak

b-kelly committed:
cleanup

b-kelly committed:
cleanup

b-kelly committed:
Update docs/product/components/partials/_topbar.html

Co-authored-by: Aaron Shekey <aaron@aaronshekey.com>

b-kelly committed:
cleanup

b-kelly committed:
cleanup docs

b-kelly committed:
add markdown filter

b-kelly committed:
cleanup

b-kelly committed:
remove unused styles

b-kelly committed:
tweak selected hover styles to display

b-kelly committed:
drop s-topbar--navigation

b-kelly committed:
quick fix for light/dark topbar navigation

b-kelly committed:
cleanup

dancormier committed:
Ensure search select vertically aligns w/ input

dancormier committed:
Ensure search input fills vertical space of parent

b-kelly committed:
Update docs/product/components/partials/_topbar.html

Co-authored-by: Aaron Shekey <aaron@aaronshekey.com>

b-kelly commented on lib/css/components/_stacks-badges.less:
```diff
@@ -20,10 +20,6 @@
     background-color: @background;
     color: @color;
-
-    .s-tag--dismiss {
```
This was not used anywhere in the product, nor in Core. Furthermore, it was bloating the compiled css file, so I deleted it.

aaronshekey replied:
Are we absolutely sure? We use something that looks just like it in the context of the tag editor. I bet if we switch to `s-tag` there, we'll miss this.

b-kelly committed:
change `--topbar-height` to `--theme-topbar-height` to align with other component variables

b-kelly committed:
cleanup

b-kelly committed:
cleanup

b-kelly committed:
scope `transition: all` to specific properties

b-kelly committed:
fix for using `.s-topbar__{light,dark}` as mixins

aaronshekey committed:
Better dropdown caret positioning

aaronshekey committed:
Better spacing

aaronshekey committed:
No wrapping on navigation items

aaronshekey committed:
Tighten up navigation spacing

aaronshekey committed:
Let's call it .s-topbar--menu-btn

I found this to be confusing at a glance. Is this the navigation? Is this the cta section? I think it's a little more clear that it's the menu button now.

aaronshekey committed:
Give the notice some breathing room

aaronshekey committed:
Remove re-importing

I'm not able to reproduce what this issue was :v:

aaronshekey committed:
Front matter and a bit of re-organization

aaronshekey committed:
Higher contrast on forced dark mode input borders

aaronshekey committed:
Comment formatting

aaronshekey committed:
Example numbers

aaronshekey committed:
Make sure our activity indicator transitions work when scrolling

When the container shows a scrollbar, this works now.

aaronshekey committed:
Consistent standalone search button styling

aaronshekey committed:
Whitespace

aaronshekey committed:
No shrinking in notices

aaronshekey committed:
Wrapping stuff

Should we scroll the CTAs?

aaronshekey reviewed (changes requested):
So very close now!

- [x] Grab the high contrast overrides we have in production and add them here
- [x] Double check the forced dark mode scrollbar colors. Transparency approach doesn't work there
- [x] Should we also apply scrolling to the CTA's section? We could probably kill it.
- [x] Navigation button stays in `x` state when navigating to a new page in docs.

aaronshekey committed:
Apostrophe

aaronshekey commented on docs/_data/topbar.json:
```diff
@@ -0,0 +1,66 @@
+    },
+    {
+        "class": ".s-topbar--ctas",
+        "description": "Basically --content above... why do I have this again? Hover classes maybe?",
```
Right, heh. This was my thought as well.

b-kelly replied:
I'll update this section. The reasoning is that we want to maintain sizing, spacing, etc on these items, but we don't want to style them any further. That way, we can add in buttons or whatever without worrying about stepping on _their_ hover classes. This could probably be renamed to something like `.s-topbar--content.s-topbar--content__unstyled`. Thoughts?

b-kelly committed:
high contrast tweaks

b-kelly committed:
fix forced mode scrollbar colors

b-kelly committed:
update hamburger menu on navigation

aaronshekey committed:
Merge branch 'develop' into bkelly/topbar-component

aaronshekey committed:
Comment refinement

aaronshekey committed:
We also need scrolling on the CTAs section

aaronshekey committed:
Add some explainers

aaronshekey committed:
Clarification of .s-topbar--ctas

b-kelly committed:
remove unnecessary safari hack

b-kelly committed:
Merge branch 'develop' into bkelly/topbar-component

b-kelly committed:
move s-navigation component to stacks-static since it uses no dynamic variables and s-topbar relies on it

aaronshekey commented:
Small regression with the navigation text color:
<img width="408" alt="image" src="https://user-images.githubusercontent.com/1369864/139135047-c995446f-4a54-4eb7-8baf-a8ccd01fc9b3.png">

We just gotta do more states here, if I'm not mistaken. Active, and is-selected:hover states.

aaronshekey commented:
<img width="113" alt="image" src="https://user-images.githubusercontent.com/1369864/139136166-87e98531-97cf-4f46-b4ec-4492dbb889b1.png">

This `new` button needs an active state. Clicking it in Safari makes it lose its background.

b-kelly committed:
don't style is-selected nav items

b-kelly commented:
> This `new` button needs an active state

Can you expand on that? I have JS added to the docs to toggle the state of the notices on click. I did this so users can easily preview what it looks like with and without the `is-unread` styling.

aaronshekey committed:
Merge branch 'develop' into bkelly/topbar-component

aaronshekey reviewed (approved):
🦐
